src/NumericalAlgorithms/Spectral/SwshCollocation.hpp

```diff
@@ -0,0 +1,199 @@
+  };
+
+  /// The representation of the block of complex values, which sets the stride
+  /// inside the libsharp type
+  static constexpr ComplexRepresentation complex_representation =
+      Representation;
+
+  /// \brief Generates the libsharp collocation information and stores it in
+  /// `geom_info_`.
+  ///
+  /// \note If you will potentially use the same `l_max` collocation set more
+  /// than once, it is probably better to use the
+  /// `precomputed_spherical_harmonic_collocation` function
+  explicit Collocation(size_t l_max) noexcept;
+
+  /// default constructor required for iterator use
+  ~Collocation() = default;
+  Collocation() = default;
+  Collocation(const Collocation&) = default;
+  Collocation(Collocation&&) = default;
+  Collocation& operator=(Collocation&) = default;
+  Collocation& operator=(Collocation&&) = default;
+
+  /// retrieve the `sharp_geom_info*` stored. This should largely be used only
+  /// for passing to other libsharp functions. Otherwise, access elements
+  /// through iterator or access functions.
+  // clang-tidy seems to think the pointer will implicitly convert to the
+  // libsharp type. It does not.
+  sharp_geom_info* get_sharp_geom_info() const noexcept {
+    return *(geom_info_.get());  // NOLINT
+  }
+
+  /// Retrieve the \f$\theta\f$ value for a given index in a libsharp-compatible
+  /// array
+  double theta(size_t offset) const noexcept;
+  /// Retrieve the \f$\phi\f$ value for a given index in a libsharp-compatible
+  /// array
+  double phi(size_t offset) const noexcept;
+
+  constexpr size_t l_max() const noexcept { return l_max_; }
+
+  /// Compute the number of entries the libsharp-compatible data structure
+  /// should have
+  constexpr size_t size() const noexcept {
+    return (l_max_ + 1) * (2 * l_max_ + 1);
+  }
+
+  // @{
+  /// Get a bidirectional iterator to the start of the grid. `operator*` for
+  /// that iterator gives a
+  /// `std::pair<size_t /*offset*/ std::pair<double /*theta*/, double /*phi*/>`
+  Collocation<Representation>::CollocationConstIterator begin() const noexcept {
+    return CollocationConstIterator{make_not_null(this), 0};
+  }
+  Collocation<Representation>::CollocationConstIterator cbegin() const
+      noexcept {
+    return begin();
+  }
+  // @}
+  // @{
+  /// Get a bidirectional iterator to the end of the grid. `operator*` for
+  /// that iterator gives a
+  /// `std::pair<size_t /*offset*/ std::pair<double /*theta*/, double /*phi*/>`
+  Collocation<Representation>::CollocationConstIterator end() const noexcept {
+    return CollocationConstIterator{make_not_null(this), size()};
+  }
+  Collocation<Representation>::CollocationConstIterator cend() const noexcept {
+    return end();
+  }
+  // @}
+
+ private:
+  // an extra pointer layer is required for the peculiar way that libsharp
+  // constructs these values.
+  size_t l_max_ = 0;
+  std::unique_ptr<sharp_geom_info*, detail::DestroySharpGeometry> geom_info_;
```

Comment:
I believe you can use just a `std::unique_ptr<sharp_geom_info, [appropriate destructor here]>`.  The `sharp_make_gauss_geom_info` function appears to be a c-style object allocator that returns by pointer.  (I think the return-by-pointer is unusual for an allocator, but I haven't used that many of these sorts of interfaces.)  Anyway, I would expect something like this to work:
```c++
sharp_geom_info* sgi;
sharp_make_gauss_geom_info(..., &sgi);
member_unique_ptr_to_sgi.reset(sgi);  // or in the member initialization with a lambda if you prefer
```